### DISCUSSION TITLE
How to rename existing fields in Models without data loss?

### DISCUSSION THREAD

**Discussion by solankimihir** [serverpod/serverpod#2932] [Q&A] [answered]
Hi everyone,
Just got started with Serverpod. 
I have defined a model with a field having an incorrect name and want to change the name from "Calorites" to "Calories", however, if I just rename it in model and run 
`Serverpod genereate`
and
`serverpod create-migration`

I get following warning:

```
 consumeal_tracker_server > serverpod generate        
✓ Generating code (5.3s)
✓ Done.

consumeal_tracker_server > serverpod create-migration
⠹ Creating migration... (0.1s)
Migration Warnings:
 • Column "calorites" of table "master_meal_item" will be dropped.
 • One or more columns are added to table "master_meal_item" which cannot be added in a table migration. The complete table will be deleted and recreated.
Migration aborted. Use --force to ignore warnings.
✗ Creating migration (0.2s)
```

**Comment by hampuslavin** [comment]
Hi @solankimihir! 👋 
We don't currently support generating renaming migrations. However, you can always edit the migration yourself to do the rename. This should work well if you are just starting out and developing locally!

Note that if/when you have a database up and running in production it's probably best to do a proper data migration if you need to rename, as you might have instances running trying to access a column that no longer exists.

**Reply by solankimihir** [reply]
Hey, thanks for the response.

> it's probably best to do a proper data migration if you need to rename,

Could you please elaborate on proper data migration? 

Should I do the following 
1. add another column calories through models
2. create migration from CLI
3. RUN a script in SQL to copy all data from Calorites column to Calories column 
4. Update the flutter (app) code to use calories column
5. Update the model again, to remove Calorites column and generate migration

**Reply by hampuslavin** [reply, marked as answer]
Basically, yes! There are some additional things to consider:
- If you had a lot of traffic and didn't want to cause any downtime, before point 3 you also want to write to both columns. Then your script to copy over data will just be a backfill up to the point where you started writing to both columns. 
  - However, if you don't have huge demands on uptime it's easiest to just do short maintenance break where you disable CRUD operations from the app and run point 3. Then you can do just as you described above.
- Old Flutter clients will still expect the previous model, so be aware that dropping the old column will be a breaking change for those clients. The easiest is to force an app upgrade, but if you have a lot of users you might want to take a more graceful approach:
  - After backfilling the new column mark the old column as `persist=false` (instead of removing it completely) in the model file to drop it. 
  - In your server you have to check if (old) clients sent something in the old field and save it in the new column.
  - When the number of old clients is low enough you can force an app upgrade and remove it from the model completely.
  
 Hope this helps!

**Comment by dishankjindal1** [comment]
Is there any way to write migration? and apply it, which helps move data around in the db?
